src/services/upower/device.hpp

```diff
@@ -0,0 +1,174 @@
+#pragma once
+
+#include <qcontainerfwd.h>
+#include <qobject.h>
+#include <qqmlintegration.h>
+#include <qtmetamacros.h>
+#include <qtypes.h>
+
+#include "../../core/doc.hpp"
+#include "../../dbus/properties.hpp"
+#include "dbus_device.h"
+
+namespace qs::service::upower {
+
+class UPowerDeviceState: public QObject {
+	Q_OBJECT;
+	QML_ELEMENT;
+	QML_SINGLETON;
+
+public:
+	enum Enum {
+		Unknown = 0,
+		Charging = 1,
+		Discharging = 2,
+		Empty = 3,
+		FullyCharged = 4,
+		PendingCharge = 5,
+		PendingDischarge = 6,
+	};
+	Q_ENUM(Enum);
+
+	Q_INVOKABLE static QString toString(UPowerDeviceState::Enum status);
+};
+
+class UPowerDeviceType: public QObject {
+	Q_OBJECT;
+	QML_ELEMENT;
+	QML_SINGLETON;
+
+public:
+	enum Enum {
+		Unknown = 0,
+		LinePower = 1,
+		Battery = 2,
+		Ups = 3,
+		Monitor = 4,
+		Mouse = 5,
+		Keyboard = 6,
+		Pda = 7,
+		Phone = 8,
+		MediaPlayer = 9,
+		Tablet = 10,
+		Computer = 11,
+		GamingInput = 12,
+		Pen = 13,
+		Touchpad = 14,
+		Modem = 15,
+		Network = 16,
+		Headset = 17,
+		Speakers = 18,
+		Headphones = 19,
+		Video = 20,
+		OtherAudio = 21,
+		RemoteControl = 22,
+		Printer = 23,
+		Scanner = 24,
+		Camera = 25,
+		Wearable = 26,
+		Toy = 27,
+		BluetoothGeneric = 28,
+	};
+	Q_ENUM(Enum);
+
+	Q_INVOKABLE static QString toString(UPowerDeviceType::Enum type);
+};
+
+///! A device exposed through the UPower system service.
+class UPowerDevice: public QObject {
+	Q_OBJECT;
+	// clang-format off
+	/// The type of power device.
+	///
+	/// > [!INFO] When checking if the device is a battery, also verify that `powerSupply` is set to `true`.
+	Q_PROPERTY(UPowerDeviceType::Enum type READ type NOTIFY typeChanged);
+	/// If the device is a power supply.
+	Q_PROPERTY(bool powerSupply READ powerSupply NOTIFY powerSupplyChanged);
+	/// Current energy level of the device in watt-hours.
+	Q_PROPERTY(qreal energy READ energy NOTIFY energyChanged);
+	/// Maximum energy capacity of the device in watt-hours
+	Q_PROPERTY(qreal energyFull READ energyFull NOTIFY energyFullChanged);
+	/// Rate of energy change in watts (positive when charging, negative when discharging).
+	Q_PROPERTY(qreal energyRate READ energyRate NOTIFY energyRateChanged);
+	/// Estimated time until the device is fully discharged, in seconds.
+	///
+	/// Will be set to `0` if charging.
+	Q_PROPERTY(qreal timeToEmpty READ timeToEmpty NOTIFY timeToEmptyChanged);
+	/// Estimated time until the device is fully charged, in seconds.
+	///
+	/// Will be set to `0` if discharging.
+	Q_PROPERTY(qreal timeToFull READ timeToFull NOTIFY timeToFullChanged);
+	/// Current charge level as a percentage.
+	Q_PROPERTY(qreal percentage READ percentage NOTIFY percentageChanged);
+	/// If the device is currently present in the system.
+	Q_PROPERTY(bool isPresent READ isPresent NOTIFY isPresentChanged);
```

Comment:
This is unclear. Is this like if you have a removable battery and took it out or a charger that is unplugged? Add some more detail, and possibly rename to `connected` if that is the case.

Reply:
This property is relevant for batteries that can be removed from the device like batteries from older laptops. I think `isPresent` or `isPresentInBay` is still more suitable as it stays more true its behavior. I've changed the documentation and added a warning to be more specific. Let me know if there is a better way to address this. 
```
/// If the power source is present in the bay or slot, required for hot-removable batteries.
///
/// [!WARNING] If the device `type` is not `Battery`, then the property will be always be `false`.
Q_PROPERTY(bool isPresent READ isPresent NOTIFY isPresentChanged);
```

Reply:
nit: doesn't really need to be a warning, just normal text is fine